{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rappels\n",
    "\n",
    "[GO ontology](https://www.ebi.ac.uk/QuickGO/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### IFB Virtual Machine Configuration\n",
    "\n",
    "* `rootDir` points to Jupyter Home Directory\n",
    "* `projectDir` points to the Git project, holds the *tsv* file\n",
    "* `libDir` points to Git project python library with files names `go.py  stat_utils.py  uniprot.py`\n",
    "* `dataDir` points to shared data folder with uniprot and GO files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "rootDir = \"//ifb/data/mydatalocal\"\n",
    "projectDir = rootDir + \"/VDB_jupyter\"\n",
    "libDir = projectDir + \"/lib\"\n",
    "dataDir= \"/ifb/data/public/teachdata/m2binf_ucbl/data\" \n",
    "\n",
    "sys.path.append(libDir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, uniprot, go\n",
    "import numpy as np\n",
    "from stat_utils import computeORA_BKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jeux de données\n",
    "* Experimental dataset as TSV\n",
    "* Go ontology as [owl](http://www.obofoundry.org/ontology/go.html)\n",
    "* Uniprot entries as XML files\n",
    "\n",
    "## Objectifs\n",
    "1. Lire et traiter des données d'entrée avec [PANDAS](https://pandas.pydata.org/)\n",
    "\n",
    "2. Représenter la distribution empirique des \"fold-change\" (**FC**)\n",
    "\n",
    "3. Evaluer un seuil de **FDR** 5%, with normal approximation\n",
    "    1. Proposer une function qui à un seuil de faux-positif retourne une valeur de **FDC**\n",
    "    2. Proposer une fonction qui appelle la précedente et retourne, en plus, une représentation graphique\n",
    "\n",
    "4. Analyse de l'enrichissement en pathway Gene Ontology\n",
    "    1. Arbre des termes d'ontologie.\n",
    "        * Visualisation de l'arbre complet\n",
    "    2. Modélisation hypergéometrique des fréquence attendues des termes GO\n",
    "        * Les protéines très abondantes présentent-elles des termes GO en quantité remarquable ?\n",
    "        * Les protéines peu abondantes présentent-elles des termes GO en quantité remarquable ?\n",
    "    3. Visualisation de l'arbre avec style des noeuds rendant compte des protéines qu'ils annotent\n",
    " \n",
    "5. Volcano plot\n",
    "    1. Scatter-Plot **FC** vs **LOG10 Adj.P-val**\n",
    "    2. Colorer les protéines très/peu abondantes\n",
    "    3. ipyWidget slider FDR permettant de changer les seuils \"d'abondance\" \n",
    "    \n",
    "6. Volcano Plot présentant les pathways\n",
    "    1. ipyWidget Slider\n",
    "    2. Coloration des points pour les protéines annotées par ce pathway \n",
    "\n",
    "## Conclusion\n",
    "* Compare wt/mutant conditions\n",
    "\n",
    "\n",
    "## Ressources\n",
    "https://www.bioconductor.org/help/course-materials/2015/SeattleApr2015/E_GeneSetEnrichment.html\n",
    "\n",
    "https://matplotlib.org/gallery.html#\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format des données expérimentales\n",
    "\n",
    "Le terme *SA*, désigne le ratio de (**S**ur) **A**bondance pour une protéines entre deux experiences.\n",
    "Les valeurs de **SA** sont stockées dans la colonne ```Log2 Corrected Abundance Ratio```. Un \"ratio\" négatif indique qu'une protéine est moins présente dans l'experience considéré par rapport à une experience de référence.\n",
    "\n",
    "* La condition experimentale de référence est un milieu de croissance bactérienne\n",
    "* La condition experimentale testée est un milieu riche en Tétracycline \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Première exploration des données"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lecture des valeurs experimentales\n",
    "Charger ```TCL_wt1.tsv``` dans une [dataframe pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).\n",
    "* Retirer les lignes présentant des aberrations numériques\n",
    "\n",
    "##### [Comment filtrer des lignes Pandas](https://cmdlinetips.com/2018/02/how-to-subset-pandas-dataframe-based-on-values-of-a-column/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Corrected Abundance ratio (1.53)</th>\n",
       "      <th>Log2 Corrected Abundance Ratio</th>\n",
       "      <th>Abundance Ratio Adj. P-Value: (127. T3 Tc WT) / (126. T0 WT)</th>\n",
       "      <th>LOG10 Adj.P-val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P75936</td>\n",
       "      <td>Basal-body rod modification protein FlgD OS=Es...</td>\n",
       "      <td>flgD</td>\n",
       "      <td>0.075816993</td>\n",
       "      <td>-3.721334942</td>\n",
       "      <td>5.494560e-05</td>\n",
       "      <td>4.260067469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P76231</td>\n",
       "      <td>Uncharacterized protein YeaC OS=Escherichia co...</td>\n",
       "      <td>yeaC</td>\n",
       "      <td>0.092810458</td>\n",
       "      <td>-3.429568818</td>\n",
       "      <td>3.510530e-04</td>\n",
       "      <td>3.45462743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P0A8S9</td>\n",
       "      <td>Flagellar transcriptional regulator FlhD OS=Es...</td>\n",
       "      <td>flhD</td>\n",
       "      <td>0.102614379</td>\n",
       "      <td>-3.284695189</td>\n",
       "      <td>2.679790e-05</td>\n",
       "      <td>4.571899347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P05706</td>\n",
       "      <td>PTS system glucitol/sorbitol-specific EIIA com...</td>\n",
       "      <td>srlB</td>\n",
       "      <td>0.108496732</td>\n",
       "      <td>-3.204276506</td>\n",
       "      <td>1.996330e-02</td>\n",
       "      <td>1.699767669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P29744</td>\n",
       "      <td>Flagellar hook-associated protein 3 OS=Escheri...</td>\n",
       "      <td>flgL</td>\n",
       "      <td>0.124183007</td>\n",
       "      <td>-3.009460329</td>\n",
       "      <td>3.674628e-02</td>\n",
       "      <td>1.434786589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P0A9A2</td>\n",
       "      <td>Bacterial non-heme ferritin-like protein OS=Es...</td>\n",
       "      <td>ftnB</td>\n",
       "      <td>0.126797386</td>\n",
       "      <td>-2.979403095</td>\n",
       "      <td>1.296589e-03</td>\n",
       "      <td>2.887197633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P75937</td>\n",
       "      <td>Flagellar hook protein FlgE OS=Escherichia col...</td>\n",
       "      <td>flgE</td>\n",
       "      <td>0.137908497</td>\n",
       "      <td>-2.858216749</td>\n",
       "      <td>7.012880e-07</td>\n",
       "      <td>6.154103778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P43533</td>\n",
       "      <td>Flagella synthesis protein FlgN OS=Escherichia...</td>\n",
       "      <td>flgN</td>\n",
       "      <td>0.150326797</td>\n",
       "      <td>-2.733825887</td>\n",
       "      <td>1.322797e-03</td>\n",
       "      <td>2.878506676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P0ABR1</td>\n",
       "      <td>DNA damage-inducible protein I OS=Escherichia ...</td>\n",
       "      <td>dinI</td>\n",
       "      <td>0.159477124</td>\n",
       "      <td>-2.6485786</td>\n",
       "      <td>4.689497e-01</td>\n",
       "      <td>0.328873724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P26608</td>\n",
       "      <td>Flagellar secretion chaperone FliS OS=Escheric...</td>\n",
       "      <td>fliS</td>\n",
       "      <td>0.162091503</td>\n",
       "      <td>-2.625119627</td>\n",
       "      <td>5.282561e-01</td>\n",
       "      <td>0.277155457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P0ABH9</td>\n",
       "      <td>ATP-dependent Clp protease ATP-binding subunit...</td>\n",
       "      <td>clpA</td>\n",
       "      <td>0.174509804</td>\n",
       "      <td>-2.518620006</td>\n",
       "      <td>7.012880e-07</td>\n",
       "      <td>6.154103778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P52614</td>\n",
       "      <td>Flagellar hook-length control protein OS=Esche...</td>\n",
       "      <td>fliK</td>\n",
       "      <td>0.196732026</td>\n",
       "      <td>-2.345696261</td>\n",
       "      <td>7.687690e-05</td>\n",
       "      <td>4.114204046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P0ACY9</td>\n",
       "      <td>Uncharacterized protein YebG OS=Escherichia co...</td>\n",
       "      <td>yebG</td>\n",
       "      <td>0.20130719</td>\n",
       "      <td>-2.312529397</td>\n",
       "      <td>6.390213e-02</td>\n",
       "      <td>1.194484636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P0C054</td>\n",
       "      <td>Small heat shock protein IbpA OS=Escherichia c...</td>\n",
       "      <td>ibpA</td>\n",
       "      <td>0.202614379</td>\n",
       "      <td>-2.303191532</td>\n",
       "      <td>8.465010e-05</td>\n",
       "      <td>4.072372404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P69741</td>\n",
       "      <td>Hydrogenase-2 small chain OS=Escherichia coli ...</td>\n",
       "      <td>hybO</td>\n",
       "      <td>0.208496732</td>\n",
       "      <td>-2.261903324</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P06959</td>\n",
       "      <td>Dihydrolipoyllysine-residue acetyltransferase ...</td>\n",
       "      <td>aceF</td>\n",
       "      <td>0.210457516</td>\n",
       "      <td>-2.248399059</td>\n",
       "      <td>1.408610e-06</td>\n",
       "      <td>5.851209497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P0AER0</td>\n",
       "      <td>Glycerol uptake facilitator protein OS=Escheri...</td>\n",
       "      <td>glpF</td>\n",
       "      <td>0.226143791</td>\n",
       "      <td>-2.14468771</td>\n",
       "      <td>2.680476e-02</td>\n",
       "      <td>1.571788011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P52613</td>\n",
       "      <td>Flagellar FliJ protein OS=Escherichia coli (st...</td>\n",
       "      <td>fliJ</td>\n",
       "      <td>0.226143791</td>\n",
       "      <td>-2.14468771</td>\n",
       "      <td>6.065791e-01</td>\n",
       "      <td>0.217112546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P0AGK8</td>\n",
       "      <td>HTH-type transcriptional regulator IscR OS=Esc...</td>\n",
       "      <td>iscR</td>\n",
       "      <td>0.226797386</td>\n",
       "      <td>-2.140524085</td>\n",
       "      <td>2.712801e-02</td>\n",
       "      <td>1.566582123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P24216</td>\n",
       "      <td>Flagellar hook-associated protein 2 OS=Escheri...</td>\n",
       "      <td>fliD</td>\n",
       "      <td>0.22745098</td>\n",
       "      <td>-2.136372442</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P0AEM4</td>\n",
       "      <td>Negative regulator of flagellin synthesis OS=E...</td>\n",
       "      <td>flgM</td>\n",
       "      <td>0.22875817</td>\n",
       "      <td>-2.128104826</td>\n",
       "      <td>5.734312e-02</td>\n",
       "      <td>1.24151868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P09373</td>\n",
       "      <td>Formate acetyltransferase 1 OS=Escherichia col...</td>\n",
       "      <td>pflB</td>\n",
       "      <td>0.232679739</td>\n",
       "      <td>-2.103582507</td>\n",
       "      <td>3.566570e-06</td>\n",
       "      <td>5.447748877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P52643</td>\n",
       "      <td>D-lactate dehydrogenase OS=Escherichia coli (s...</td>\n",
       "      <td>ldhA</td>\n",
       "      <td>0.232679739</td>\n",
       "      <td>-2.103582507</td>\n",
       "      <td>2.699170e-03</td>\n",
       "      <td>2.5687698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P04982</td>\n",
       "      <td>D-ribose pyranase OS=Escherichia coli (strain ...</td>\n",
       "      <td>rbsD</td>\n",
       "      <td>0.237254902</td>\n",
       "      <td>-2.0754902</td>\n",
       "      <td>1.238910e-04</td>\n",
       "      <td>3.906960599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P0AFP6</td>\n",
       "      <td>GTP cyclohydrolase 1 type 2 homolog OS=Escheri...</td>\n",
       "      <td>ybgI</td>\n",
       "      <td>0.237254902</td>\n",
       "      <td>-2.0754902</td>\n",
       "      <td>1.386319e-01</td>\n",
       "      <td>0.85813682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P0A9G6</td>\n",
       "      <td>Isocitrate lyase OS=Escherichia coli (strain K...</td>\n",
       "      <td>aceA</td>\n",
       "      <td>0.238562092</td>\n",
       "      <td>-2.067563284</td>\n",
       "      <td>7.901680e-07</td>\n",
       "      <td>6.102280415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P33602</td>\n",
       "      <td>NADH-quinone oxidoreductase subunit G OS=Esche...</td>\n",
       "      <td>nuoG</td>\n",
       "      <td>0.250980392</td>\n",
       "      <td>-1.994353437</td>\n",
       "      <td>4.319700e-07</td>\n",
       "      <td>6.364546423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P0A759</td>\n",
       "      <td>Glucosamine-6-phosphate deaminase OS=Escherich...</td>\n",
       "      <td>nagB</td>\n",
       "      <td>0.253594771</td>\n",
       "      <td>-1.979403095</td>\n",
       "      <td>1.396989e-01</td>\n",
       "      <td>0.854806917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P41409</td>\n",
       "      <td>Pyrimidine-specific ribonucleoside hydrolase R...</td>\n",
       "      <td>rihA</td>\n",
       "      <td>0.255555556</td>\n",
       "      <td>-1.96829114</td>\n",
       "      <td>7.742950e-04</td>\n",
       "      <td>3.111093695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>P37646</td>\n",
       "      <td>Cyclic di-GMP phosphodiesterase PdeH OS=Escher...</td>\n",
       "      <td>yhjH</td>\n",
       "      <td>0.25620915</td>\n",
       "      <td>-1.964606093</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1716</th>\n",
       "      <td>P0ADB7</td>\n",
       "      <td>Entericidin B OS=Escherichia coli (strain K12)...</td>\n",
       "      <td>ecnB</td>\n",
       "      <td>1.077124183</td>\n",
       "      <td>0.10718459</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1717</th>\n",
       "      <td>P0A915</td>\n",
       "      <td>Outer membrane protein W OS=Escherichia coli (...</td>\n",
       "      <td>ompW</td>\n",
       "      <td>1.078431373</td>\n",
       "      <td>0.108934372</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1718</th>\n",
       "      <td>P39180</td>\n",
       "      <td>Antigen 43 OS=Escherichia coli (strain K12) OX...</td>\n",
       "      <td>flu</td>\n",
       "      <td>1.082352941</td>\n",
       "      <td>0.11417102</td>\n",
       "      <td>2.560931e-03</td>\n",
       "      <td>2.591602196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1719</th>\n",
       "      <td>P0ACF8</td>\n",
       "      <td>DNA-binding protein H-NS OS=Escherichia coli (...</td>\n",
       "      <td>hns</td>\n",
       "      <td>1.083006536</td>\n",
       "      <td>0.11504195</td>\n",
       "      <td>2.679790e-05</td>\n",
       "      <td>4.571899347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1720</th>\n",
       "      <td>P0A7K2</td>\n",
       "      <td>50S ribosomal protein L7/L12 OS=Escherichia co...</td>\n",
       "      <td>rplL</td>\n",
       "      <td>1.094117647</td>\n",
       "      <td>0.129767875</td>\n",
       "      <td>2.483148e-02</td>\n",
       "      <td>1.604997385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1721</th>\n",
       "      <td>P0A7G2</td>\n",
       "      <td>30S ribosome-binding factor OS=Escherichia col...</td>\n",
       "      <td>rbfA</td>\n",
       "      <td>1.095424837</td>\n",
       "      <td>0.131490496</td>\n",
       "      <td>1.356704e-01</td>\n",
       "      <td>0.867514879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1722</th>\n",
       "      <td>P0A937</td>\n",
       "      <td>Outer membrane protein assembly factor BamE OS...</td>\n",
       "      <td>bamE</td>\n",
       "      <td>1.123529412</td>\n",
       "      <td>0.168037892</td>\n",
       "      <td>1.294795e-02</td>\n",
       "      <td>1.887799128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1723</th>\n",
       "      <td>P0A9L8</td>\n",
       "      <td>Pyrroline-5-carboxylate reductase OS=Escherich...</td>\n",
       "      <td>proC</td>\n",
       "      <td>1.133333333</td>\n",
       "      <td>0.180572246</td>\n",
       "      <td>9.350732e-01</td>\n",
       "      <td>0.029154378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1724</th>\n",
       "      <td>P62399</td>\n",
       "      <td>50S ribosomal protein L5 OS=Escherichia coli (...</td>\n",
       "      <td>C813_RS16705; C813_RS25010; rplE; AB182_RS27795</td>\n",
       "      <td>1.133986928</td>\n",
       "      <td>0.18140401</td>\n",
       "      <td>5.518020e-06</td>\n",
       "      <td>5.258216386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1725</th>\n",
       "      <td>P0AER8</td>\n",
       "      <td>Sodium/glutamate symporter OS=Escherichia coli...</td>\n",
       "      <td>gltS</td>\n",
       "      <td>1.138562092</td>\n",
       "      <td>0.187212971</td>\n",
       "      <td>2.246621e-01</td>\n",
       "      <td>0.648470173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1726</th>\n",
       "      <td>P0A905</td>\n",
       "      <td>Outer membrane lipoprotein SlyB OS=Escherichia...</td>\n",
       "      <td>slyB</td>\n",
       "      <td>1.173856209</td>\n",
       "      <td>0.231255697</td>\n",
       "      <td>5.109990e-05</td>\n",
       "      <td>4.29157999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1727</th>\n",
       "      <td>P76471</td>\n",
       "      <td>Outer membrane protein YfaZ OS=Escherichia col...</td>\n",
       "      <td>yfaZ</td>\n",
       "      <td>1.178431373</td>\n",
       "      <td>0.236867744</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1728</th>\n",
       "      <td>P76506</td>\n",
       "      <td>Probable phospholipid-binding lipoprotein MlaA...</td>\n",
       "      <td>mlaA</td>\n",
       "      <td>1.21372549</td>\n",
       "      <td>0.279442162</td>\n",
       "      <td>1.408320e-05</td>\n",
       "      <td>4.851297862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1729</th>\n",
       "      <td>P13036</td>\n",
       "      <td>Fe(3+) dicitrate transport protein FecA OS=Esc...</td>\n",
       "      <td>fecA</td>\n",
       "      <td>1.230718954</td>\n",
       "      <td>0.299501347</td>\n",
       "      <td>2.679790e-05</td>\n",
       "      <td>4.571899347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730</th>\n",
       "      <td>P0A927</td>\n",
       "      <td>Nucleoside-specific channel-forming protein Ts...</td>\n",
       "      <td>tsx</td>\n",
       "      <td>1.254901961</td>\n",
       "      <td>0.327574658</td>\n",
       "      <td>1.397352e-02</td>\n",
       "      <td>1.85469412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1731</th>\n",
       "      <td>P0AB46</td>\n",
       "      <td>Uncharacterized protein YmgD OS=Escherichia co...</td>\n",
       "      <td>ymgD</td>\n",
       "      <td>1.261437908</td>\n",
       "      <td>0.335069195</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1732</th>\n",
       "      <td>P10384</td>\n",
       "      <td>Long-chain fatty acid transport protein OS=Esc...</td>\n",
       "      <td>fadL</td>\n",
       "      <td>1.283660131</td>\n",
       "      <td>0.360263277</td>\n",
       "      <td>1.849300e-04</td>\n",
       "      <td>3.732993373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733</th>\n",
       "      <td>P06971</td>\n",
       "      <td>Ferrichrome outer membrane transporter/phage r...</td>\n",
       "      <td>fhuA</td>\n",
       "      <td>1.293464052</td>\n",
       "      <td>0.371239959</td>\n",
       "      <td>6.974880e-06</td>\n",
       "      <td>5.156463067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1734</th>\n",
       "      <td>P0A910</td>\n",
       "      <td>Outer membrane protein A OS=Escherichia coli (...</td>\n",
       "      <td>ompA</td>\n",
       "      <td>1.294771242</td>\n",
       "      <td>0.372697227</td>\n",
       "      <td>7.838330e-06</td>\n",
       "      <td>5.105776203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1735</th>\n",
       "      <td>P0A908</td>\n",
       "      <td>MltA-interacting protein OS=Escherichia coli (...</td>\n",
       "      <td>mipA</td>\n",
       "      <td>1.335294118</td>\n",
       "      <td>0.417157551</td>\n",
       "      <td>1.606878e-02</td>\n",
       "      <td>1.794017208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1736</th>\n",
       "      <td>P75780</td>\n",
       "      <td>Catecholate siderophore receptor Fiu OS=Escher...</td>\n",
       "      <td>fiu</td>\n",
       "      <td>1.349019608</td>\n",
       "      <td>0.431911318</td>\n",
       "      <td>9.682374e-02</td>\n",
       "      <td>1.014018129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>P06996</td>\n",
       "      <td>Outer membrane protein C OS=Escherichia coli (...</td>\n",
       "      <td>ompC</td>\n",
       "      <td>1.377124183</td>\n",
       "      <td>0.461658661</td>\n",
       "      <td>8.100070e-07</td>\n",
       "      <td>6.091511321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>P46474</td>\n",
       "      <td>Uncharacterized protein YhdP OS=Escherichia co...</td>\n",
       "      <td>yhdP</td>\n",
       "      <td>1.465359477</td>\n",
       "      <td>0.551254625</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>P76344</td>\n",
       "      <td>Metal-binding protein ZinT OS=Escherichia coli...</td>\n",
       "      <td>zinT</td>\n",
       "      <td>1.503267974</td>\n",
       "      <td>0.588102208</td>\n",
       "      <td>1.236300e-05</td>\n",
       "      <td>4.907875784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1740</th>\n",
       "      <td>P06129</td>\n",
       "      <td>Vitamin B12 transporter BtuB OS=Escherichia co...</td>\n",
       "      <td>btuB</td>\n",
       "      <td>1.520915033</td>\n",
       "      <td>0.604939558</td>\n",
       "      <td>3.964344e-03</td>\n",
       "      <td>2.401828628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1741</th>\n",
       "      <td>P77330</td>\n",
       "      <td>Prophage lipoprotein Bor homolog OS=Escherichi...</td>\n",
       "      <td>borD</td>\n",
       "      <td>1.535947712</td>\n",
       "      <td>0.619129104</td>\n",
       "      <td>3.107254e-01</td>\n",
       "      <td>0.507623276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1742</th>\n",
       "      <td>P02930</td>\n",
       "      <td>Outer membrane protein TolC OS=Escherichia col...</td>\n",
       "      <td>tolC</td>\n",
       "      <td>1.552287582</td>\n",
       "      <td>0.634395861</td>\n",
       "      <td>1.337345e-02</td>\n",
       "      <td>1.873756665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1743</th>\n",
       "      <td>P0A917</td>\n",
       "      <td>Outer membrane protein X OS=Escherichia coli (...</td>\n",
       "      <td>ompX</td>\n",
       "      <td>1.579738562</td>\n",
       "      <td>0.65968582</td>\n",
       "      <td>2.226431e-03</td>\n",
       "      <td>2.652390664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>P02931</td>\n",
       "      <td>Outer membrane protein F OS=Escherichia coli (...</td>\n",
       "      <td>ompF</td>\n",
       "      <td>1.754901961</td>\n",
       "      <td>0.811390435</td>\n",
       "      <td>6.839810e-05</td>\n",
       "      <td>4.16495627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1745</th>\n",
       "      <td>P0AB40</td>\n",
       "      <td>Multiple stress resistance protein BhsA OS=Esc...</td>\n",
       "      <td>bhsA</td>\n",
       "      <td>1.798039216</td>\n",
       "      <td>0.846424487</td>\n",
       "      <td>3.592849e-02</td>\n",
       "      <td>1.444561032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1746 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Accession                                        Description Gene Symbol  \\\n",
       "0       P75936  Basal-body rod modification protein FlgD OS=Es...        flgD   \n",
       "1       P76231  Uncharacterized protein YeaC OS=Escherichia co...        yeaC   \n",
       "2       P0A8S9  Flagellar transcriptional regulator FlhD OS=Es...        flhD   \n",
       "3       P05706  PTS system glucitol/sorbitol-specific EIIA com...        srlB   \n",
       "4       P29744  Flagellar hook-associated protein 3 OS=Escheri...        flgL   \n",
       "...        ...                                                ...         ...   \n",
       "1741    P77330  Prophage lipoprotein Bor homolog OS=Escherichi...        borD   \n",
       "1742    P02930  Outer membrane protein TolC OS=Escherichia col...        tolC   \n",
       "1743    P0A917  Outer membrane protein X OS=Escherichia coli (...        ompX   \n",
       "1744    P02931  Outer membrane protein F OS=Escherichia coli (...        ompF   \n",
       "1745    P0AB40  Multiple stress resistance protein BhsA OS=Esc...        bhsA   \n",
       "\n",
       "     Corrected Abundance ratio (1.53) Log2 Corrected Abundance Ratio  \\\n",
       "0                         0.075816993                   -3.721334942   \n",
       "1                         0.092810458                   -3.429568818   \n",
       "2                         0.102614379                   -3.284695189   \n",
       "3                         0.108496732                   -3.204276506   \n",
       "4                         0.124183007                   -3.009460329   \n",
       "...                               ...                            ...   \n",
       "1741                      1.535947712                    0.619129104   \n",
       "1742                      1.552287582                    0.634395861   \n",
       "1743                      1.579738562                     0.65968582   \n",
       "1744                      1.754901961                    0.811390435   \n",
       "1745                      1.798039216                    0.846424487   \n",
       "\n",
       "      Abundance Ratio Adj. P-Value: (127. T3 Tc WT) / (126. T0 WT)  \\\n",
       "0                                              0.000055              \n",
       "1                                              0.000351              \n",
       "2                                              0.000027              \n",
       "3                                              0.019963              \n",
       "4                                              0.036746              \n",
       "...                                                 ...              \n",
       "1741                                           0.310725              \n",
       "1742                                           0.013373              \n",
       "1743                                           0.002226              \n",
       "1744                                           0.000068              \n",
       "1745                                           0.035928              \n",
       "\n",
       "     LOG10 Adj.P-val  \n",
       "0        4.260067469  \n",
       "1         3.45462743  \n",
       "2        4.571899347  \n",
       "3        1.699767669  \n",
       "4        1.434786589  \n",
       "...              ...  \n",
       "1741     0.507623276  \n",
       "1742     1.873756665  \n",
       "1743     2.652390664  \n",
       "1744      4.16495627  \n",
       "1745     1.444561032  \n",
       "\n",
       "[1746 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas\n",
    "df=pandas.read_csv(projectDir + \"/TCL_wt1.tsv\", delimiter =\"\\t\")\n",
    "df.head()\n",
    "df1 = df.dropna()\n",
    "df1.reset_index(drop=True)\n",
    "#newdf = df[ (df[\"Log2 Corrected Abundance Ratio\"] != '#VALEUR!') & (df['LOG10 Adj.P-val'] != '#VALEUR!')]\n",
    "#newdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualisation de la distribution empirique\n",
    "Représenter la distribution des **SA** de l'experience.\n",
    "* [matplotlib histogramme](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.hist.html)\n",
    "* implémenter ce graphique dans une fonction \n",
    "```python \n",
    "    def empirical(data:pandas.Series, bins:Number) -> (fig,ax)\n",
    "```\n",
    "##### [API des colonnes pandas](https://pandas.pydata.org/pandas-docs/version/0.24/reference/series.html)\n",
    "\n",
    "#### Pour aller plus loin\n",
    "Réaliser une Kernel Density Estimation sur un histogramme matplotlib doit être fait à l'aide du package scipy.\n",
    "Heureusement le package seaborn propose une macro pour cela.\n",
    "\n",
    "`seaborn.kdeplot(np.array(data), bw=0.5)`\n",
    "* implémenter ce graphique dans une fonction \n",
    "```python \n",
    "    def empiricalKDE(data:pandas.Series, bins:Number, bandwidth:Number) -> (fig,ax)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_df1 = list(df1[\"Log2 Corrected Abundance Ratio\"])\n",
    "data_df1=list(map(float,list_df1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAPOElEQVR4nO3dYaxkZX3H8e+vULFqjJC90JWFXtpsbFdjq7lBW5OGZiVsxbC0KcmSaDeVZmOCrW3ayFISedFssolNq2lqk41QNimFblDDRlLLdltDmhR1UVSWBdkIwsrKXiXWphrs6r8v7uCO17ncO3Nm7tz77Pfz5sx5zjkz/5zM/c1znznPmVQVkqS2/My0C5AkjZ/hLkkNMtwlqUGGuyQ1yHCXpAadO+0CADZs2FCzs7PTLkOS1pWHHnroW1U1M2jbmgj32dlZjhw5Mu0yJGldSfL1pbY5LCNJDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ1aEzNUJa0Ps7vvG9j+1N6rV7kSLceeuyQ1aNlwT3J7klNJHhmw7c+TVJINfW03Jzme5PEkV427YEnS8lbSc78D2La4McklwJXA031tW4AdwOt7x3w0yTljqVSStGLLhntVPQA8P2DT3wAfAPp/YXs7cHdVvVBVTwLHgcvHUagkaeVGGnNPcg3wjar60qJNFwPP9K2f6LUNeo5dSY4kOTI/Pz9KGZKkJQwd7kleAdwCfHDQ5gFtNaCNqtpXVXNVNTczM/Be85KkEY1yKeQvAZcBX0oCsAn4QpLLWeipX9K37ybg2a5FSpKGM3TPvaq+UlUXVtVsVc2yEOhvrqpvAgeBHUnOS3IZsBn43FgrliQtayWXQt4F/BfwuiQnktyw1L5VdRQ4ADwKfBq4sap+OK5iJUkrs+ywTFVdv8z22UXre4A93cqSJHXhDFVJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQcv+hqokLWd2930/fvzU3qunWIletGzPPcntSU4leaSv7UNJHkvy5SSfTPKavm03Jzme5PEkV02qcEnS0lYyLHMHsG1R2yHgDVX1RuCrwM0ASbYAO4DX9475aJJzxlatJGlFlg33qnoAeH5R2/1Vdbq3+iCwqfd4O3B3Vb1QVU8Cx4HLx1ivJGkFxvGF6nuAf+k9vhh4pm/biV7bT0myK8mRJEfm5+fHUIYk6UWdwj3JLcBp4M4XmwbsVoOOrap9VTVXVXMzMzNdypAkLTLy1TJJdgLvBLZW1YsBfgK4pG+3TcCzo5cnSRrFSD33JNuAm4Brqup7fZsOAjuSnJfkMmAz8LnuZUqShrFszz3JXcAVwIYkJ4BbWbg65jzgUBKAB6vqvVV1NMkB4FEWhmturKofTqp4SdJgy4Z7VV0/oPm2l9h/D7CnS1GSpG6coSrpJfXPPtX64b1lJKlBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYtG+5Jbk9yKskjfW0XJDmU5Ine8vy+bTcnOZ7k8SRXTapwSdLSVtJzvwPYtqhtN3C4qjYDh3vrJNkC7ABe3zvmo0nOGVu1kqQVWTbcq+oB4PlFzduB/b3H+4Fr+9rvrqoXqupJ4Dhw+ZhqlSSt0Khj7hdV1UmA3vLCXvvFwDN9+53otf2UJLuSHElyZH5+fsQyJEmDjPsL1Qxoq0E7VtW+qpqrqrmZmZkxlyFJZ7dRw/25JBsBestTvfYTwCV9+20Cnh29PEnSKEYN94PAzt7jncC9fe07kpyX5DJgM/C5biVKkoZ17nI7JLkLuALYkOQEcCuwFziQ5AbgaeA6gKo6muQA8ChwGrixqn44odolTcjs7vumXYI6Wjbcq+r6JTZtXWL/PcCeLkVJkrpxhqokNchwl6QGGe6S1CDDXZIatOwXqpLODl4h0xZ77pLUIMNdkhpkuEtSgwx3SWqQ4S5JDfJqGUlj1X/VzVN7r55iJWc3e+6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktSgTuGe5E+THE3ySJK7krw8yQVJDiV5orc8f1zFSpJWZuTbDyS5GPhjYEtVfT/JAWAHsAU4XFV7k+wGdgM3jaVaSWPlD3S0q+uwzLnAzyU5F3gF8CywHdjf274fuLbja0iShjRyz72qvpHkr4Cnge8D91fV/UkuqqqTvX1OJrlw0PFJdgG7AC699NJRy5A0BHvqZ4+Re+69sfTtwGXAa4FXJnnXSo+vqn1VNVdVczMzM6OWIUkaoMuwzNuBJ6tqvqr+D/gE8BvAc0k2AvSWp7qXKUkaRpdwfxp4a5JXJAmwFTgGHAR29vbZCdzbrURJ0rC6jLl/Nsk9wBeA08AXgX3Aq4ADSW5g4QPgunEUKklauU6/xFRVtwK3Lmp+gYVevCRpSpyhKkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDWoU7gneU2Se5I8luRYkl9PckGSQ0me6C3PH1exkqSV6dpz/wjw6ar6ZeBXgWPAbuBwVW0GDvfWJUmraORwT/Jq4DeB2wCq6gdV9R1gO7C/t9t+4NquRUqShtOl5/6LwDzwD0m+mORjSV4JXFRVJwF6ywvHUKckaQhdwv1c4M3A31fVm4D/ZYghmCS7khxJcmR+fr5DGZKkxc7tcOwJ4ERVfba3fg8L4f5cko1VdTLJRuDUoIOrah+wD2Bubq461CEJmN19348fP7X36ilWorVg5J57VX0TeCbJ63pNW4FHgYPAzl7bTuDeThVKkobWpecO8EfAnUleBnwN+AMWPjAOJLkBeBq4ruNrSJKG1Cncq+phYG7Apq1dnleS1I0zVCWpQYa7JDWo65i7JC3JK3imx567JDXIcJekBjksIzWuf2hEZw977pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CBnqErrjDfj0krYc5ekBhnuktQgw12SGmS4S1KDOod7knOSfDHJp3rrFyQ5lOSJ3vL87mVKkoYxjp77+4Fjfeu7gcNVtRk43FuXJK2iTuGeZBNwNfCxvubtwP7e4/3AtV1eQ5I0vK499w8DHwB+1Nd2UVWdBOgtLxx0YJJdSY4kOTI/P9+xDElSv5HDPck7gVNV9dAox1fVvqqaq6q5mZmZUcuQJA3QZYbq24BrkrwDeDnw6iT/CDyXZGNVnUyyETg1jkIlSSs3cs+9qm6uqk1VNQvsAP69qt4FHAR29nbbCdzbuUpJ0lAmcZ37XuDKJE8AV/bWJUmraCw3DquqzwCf6T3+NrB1HM8raTT9NxfT2ckZqpLUIG/5K61R3tpXXdhzl6QG2XOX1gHH0DUsw11ax9ZT6DvMtLoclpGkBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yElM0hqyniYlaW0z3KVV4gxNrSaHZSSpQYa7JDXIcJekBhnuktQgv1CVpswrZDQJI/fck1yS5D+SHEtyNMn7e+0XJDmU5Ine8vzxlStJWokuwzKngT+rql8B3grcmGQLsBs4XFWbgcO9dUnSKhp5WKaqTgIne4//J8kx4GJgO3BFb7f9wGeAmzpVKa1xw17D7lCMJm0sY+5JZoE3AZ8FLuoFP1V1MsmFSxyzC9gFcOmll46jDEnrhBO6Jq/z1TJJXgV8HPiTqvruSo+rqn1VNVdVczMzM13LkCT16RTuSX6WhWC/s6o+0Wt+LsnG3vaNwKluJUqShjXysEySALcBx6rqr/s2HQR2Ant7y3s7VSitY46ta1q6jLm/DXg38JUkD/fa/oKFUD+Q5AbgaeC6biVK64uBrrWgy9Uy/wlkic1bR31eSVJ33n5AkhpkuEtSgwx3SWqQNw6TluGEG61H9twlqUGGuyQ1yHCXpAY55i4NwfH3yfL8jo/hrma81MxQg2LtckbvZDgsI0kNsueus8KwvUN/cGNtcbhmeIa71iz/oM9ufnh247CMJDXInrvWFHtrWo7/0a2M4S4N4IeM1juHZSSpQfbcNVHj+hd6qeexhy0NZrhLaoJj8T/JcNdU+IeoaVjqP72u78G1+H52zF2SGjSxnnuSbcBHgHOAj1XV3km9lgabRG+iS89nXOPjjrPrRUu9F4Z97y9+nkm/n1ejpz+RcE9yDvB3wJXACeDzSQ5W1aOTeL1Jf2m3kn26vplGfc5R3pRLHT/pc7eSY6VxG+X91cIX+JMalrkcOF5VX6uqHwB3A9sn9FqSpEVSVeN/0uT3gG1V9Ye99XcDb6mq9/XtswvY1Vt9HfD42AtZPzYA35p2EWuA5+EMz8UZnosFg87DL1TVzKCdJzXmngFtP/EpUlX7gH0Tev11JcmRqpqbdh3T5nk4w3NxhudiwbDnYVLDMieAS/rWNwHPTui1JEmLTCrcPw9sTnJZkpcBO4CDE3otSdIiExmWqarTSd4H/CsLl0LeXlVHJ/FajXB4aoHn4QzPxRmeiwVDnYeJfKEqSZouZ6hKUoMMd0lqkOG+BiT5yyRfTvJwkvuTvHbaNU1Lkg8leax3Pj6Z5DXTrmlaklyX5GiSHyU56y4FTLItyeNJjifZPe16piXJ7UlOJXlkmOMM97XhQ1X1xqr6NeBTwAenXdAUHQLeUFVvBL4K3DzleqbpEeB3gQemXchq67uFyW8DW4Drk2yZblVTcwewbdiDDPc1oKq+27f6ShZN+DqbVNX9VXW6t/ogC3MkzkpVdayqztaZ297CpKeqHgCeH/Y47+e+RiTZA/w+8N/Ab025nLXiPcA/T7sITcXFwDN96yeAt0yplnXJcF8lSf4N+PkBm26pqnur6hbgliQ3A+8Dbl3VAlfRcueit88twGngztWsbbWt5FycpZa9hYlemuG+Sqrq7Svc9Z+A+2g43Jc7F0l2Au8EtlbjEzGGeF+cbbyFSUeOua8BSTb3rV4DPDatWqat9yMvNwHXVNX3pl2PpsZbmHTkDNU1IMnHWbjt8Y+ArwPvrapvTLeq6UhyHDgP+Hav6cGqeu8US5qaJL8D/C0wA3wHeLiqrppuVasnyTuAD3PmFiZ7plzSVCS5C7iChVv+PgfcWlW3LXuc4S5J7XFYRpIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBv0/Dy8HjyWYwPUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "def empirical(data,bins):\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot()\n",
    "    \n",
    "    ax.hist(data,bins)\n",
    "\n",
    "    return fig, ax\n",
    "\n",
    "fig, ax = empirical(data_df1,100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEICAYAAACktLTqAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAdEElEQVR4nO3dfbRcVZnn8e+PhJeBwJCQ8JYELjKRNxuw+xJkMSq9ghKBJYyIE6bF2I0GesgMvYY1TdBxYPWQNvRMO+BC6M7ISxwRjBIgBm0FJDBMK3ADtBgSNM1b7iQml5AQIogGnvlj72sqRdW9dXPr9dTvs1atW2efU+fsqvvUU/vsc/Y5igjMzKxYdmt1BczMrP6c3M3MCsjJ3cysgJzczcwKyMndzKyAnNzNzArIyd3MrICc3OtM0lxJfZLeknRb2bweSSFpW8njSw2ow3JJn6v3eq27SfqmpPWStkr6RXmMSZohabWkNyQ9JOnwBtTBsV0jJ/f6WwdcA9wyxDL7R8S4/PhvTaqX2Wh9GeiJiP2AjwPXSPojAEkTgSXAl4AJQB/w7VZV1Jzc6y4ilkTEPcCm0a5L0gRJt0paJ2mzpHty+XhJyyQN5PJlkqbkefOBDwI35D2DG3L50ZLul/SqpOckfapkOwdI+l5ukT0h6RpJj462/lYsEbEyIt4anMyPI/P0J4CVEfGdiPgNcDVwgqSjK63Lsd14Tu6t8ZKk/hzcE4dY7n8DewPHAQcC/zOX7wbcChwOHAa8CdwAEBFfBP4PMDfvGcyVtA9wP/CtvJ4LgBslHZfX9zXg18DBwOz8MHsXSTdKegNYDawHvp9nHQf80+ByEfFr4J9zeSWO7QZzcm+uV4CTSIH7R8C+wO2VFpR0CPAx4JKI2BwRv4uIhwEiYlNE3BURb0TE68B84MNDbPds4MWIuDUitkfEk8BdwCcljQHOA67K63sWWFSft2tFExH/nhS3HyR1wwy25McBr5Ut/lpedieO7eYY2+oKdJOI2EbqiwTYIGkusF7SfhGxtWzxqcCrEbG5fD2S9ia1dGYC43PxvpLGRMTbFTZ9OHCypC0lZWNJradJ+fnaknmlz812kmPsUUmfBv4c+CqwDdivbNH9gNcrrMKx3QRuubfW4CU5VWHeWmCCpP0rzLscOAo4OR/c+lDZesov9bkWeDgi9i95jIuIPwcGgO3AlJLlp+7Ce7HuM5Ydfe4rgRMGZ+TukiNzeTnHdhM4udeZpLGS9gLGAGMk7SVpbJ53sqSjJO0m6QBSi2d5RJTvzhIR64EfkPoPx0vaXdJgoO9L6ovcImkCcFXZyzcA7ymZXga8V9KFeT27SzpJ0jG5NbQEuFrS3vkA2Gfq9oFYIUg6UNIsSeMkjZF0Bql/+8d5kbuB90k6L8f/fwV+FhGry9fl2G6SiPCjjg/SWQJR9rg6z7sAeIF0gGc98A3g4CHWNYHUR7gB2AwsyeWHAstJu8K/AC7O2xmb55+SyzcDX81lRwH3kVozm0hfyhPzvEl53lbgCeBa4MFWf5Z+tM8jx8jDwJYcJ88Any9b5nTSgdY3c3z2DLE+x3aDH8ofgNnvSbqW9KPTNWcWWHfopth2t4wNnid8vJLpwEWk3WyzjtbNse2zZQxSP+cdpF3ijcDfAve2tEZm9dG1se1uGTOzAnK3jJlZAbVFt8zEiROjp6en1dWwAluxYsUrETGp2dt1bFsjDRXXbZHce3p66OvrG35Bs10k6aVWbNexbY00VFy7W8bMrICc3K2r5dGWT0lalqcn5MvH/jL/HV+y7JWS1uTLyp7RulqbDc/J3brdZcCqkul5pBGM04AH8zSSjgVmkS5RO5M0dH5Mk+tqVrO26HO39tEz776alntxwVkNrknj5ZtAnEW6rOx/ysXnAKfl54tIQ+GvyOV3RrpZxQuS1gDTgZ80scpWZ+XxXoS4HuSWu3Wz64C/BN4pKTso0oWtyH8PzOWT2flysf257F0kzVG6j27fwMBA/WttVgMnd+tKks4GNkbEilpfUqGs4gjAiFgYEb0R0TtpUtPPvjQD3C1j3etU4OOSzgT2AvaT9E3STVQOiYj1+Y5BG/Py/ex8LfAppJuhm7Ult9ytK0XElRExJSJ6SAdKfxwRnwaWsuM+m7PZcR2SpcAsSXtKOgKYBjze5Gqb1cwtd7OdLQAWS7oIeBk4HyAiVkpaDDxLurvPpVH5tm9mbcHJ3bpeRCwnnRVDRGwCZlRZbj7pzBqztuduGTOzAqopuUvaX9J3Ja2WtErSKR7JZ2bWvmptuV8P/ENEHE26w/kqPJLPzKxtDZvcJe0HfAi4GSAifhsRW0gj9hblxRYB5+bnvx/JFxEvAIMj+czMrElqabm/h3RX8VvzBZa+LmkfRjmSz6P4zMwap5bkPhb4Q+CmiHg/8GtyF0wVNY3k8yg+M7PGqeVUyH6gPyIey9PfJSV3j+Qzs0KpdOG8Tr2Y2LAt94j4FbBW0lG5aAZpIIdH8pmZtalaBzH9B+B2SXsAzwN/Svph8Eg+M7M2VFNyj4ingd4KszySz8ysDXmEqplZATm5m5kVkJO7mVkBObmbmRWQk7uZWQE5uZuZFZCTu5lZATm5m5kVkJO7mVkBObmbmRWQk7uZWQE5uZuZFVCtV4U0M+tola7VXmRuuZuZFZCTu5lZATm5m5kVkJO7mVkBObmbmRWQz5axXVLLmQedetd4syKoueUuaYykpyQty9MTJN0v6Zf57/iSZa+UtEbSc5LOaETFzcysupF0y1wGrCqZngc8GBHTgAfzNJKOBWYBxwEzgRsljalPdc3MrBY1JXdJU4CzgK+XFJ8DLMrPFwHnlpTfGRFvRcQLwBpgen2qa2Zmtai15X4d8JfAOyVlB0XEeoD898BcPhlYW7Jcfy7biaQ5kvok9Q0MDIy44mZmVt2wyV3S2cDGiFhR4zpVoSzeVRCxMCJ6I6J30qRJNa7azMxqUcvZMqcCH5d0JrAXsJ+kbwIbJB0SEeslHQJszMv3A1NLXj8FWFfPSpuZ2dCGbblHxJURMSUiekgHSn8cEZ8GlgKz82KzgXvz86XALEl7SjoCmAY8Xveam5lZVaM5z30BsFjSRcDLwPkAEbFS0mLgWWA7cGlEvD3qmpqZWc1GNEI1IpZHxNn5+aaImBER0/LfV0uWmx8RR0bEURHxg3pX2qweJE2V9JCkVZJWSrosl3sMh3U8X37Autl24PKIOAb4AHBpHqfhMRzW8ZzcrWtFxPqIeDI/f500SG8yHsNhBeDkbgZI6gHeDzyGx3BYATi5W9eTNA64C/iLiNg61KIVyjyGw9qSk7t1NUm7kxL77RGxJBdvyGM38BgO61RO7ta1JAm4GVgVEV8pmeUxHNbxfD1362anAhcCz0h6Opd9AY/hsAJwcreuFRGPUrkfHWBGldfMB+Y3rFJmdeJuGTOzAnJyNzMrICd3M7MCcp97F6nlptZmVgxuuZuZFZCTu5lZATm5m5kVkJO7mVkB+YCqmRWOTx5wy93MrJCGTe6+FZmZWeeppVtm8FZkT0raF1gh6X7gs6RbkS2QNI90K7Irym5FdijwgKT3+gJLZtYo7oZ5t2Fb7r4VmZlZ5xlRn3s9b0VmZmaNU3Nyr/etyHyfSTOzxqkpuTfiVmS+z6SZWePUcraMb0VmZtZhajlbxrciMzPrMMMmd9+KzMys83iEqplZAfnaMgXhQRxmVsotdzOzAnLL3cw6ivdSa+PkbmY2hEo/Ji8uOKsFNRkZd8uYmRWQW+5m1tbcDbNr3HI3MysgJ3czswJycjczKyAndzOzAnJyNzMrICd3M7MC8qmQHcCngpnZSDm5m1nbcEOmfpzczawpyhN3Jwzhr6YT3ouTe4sVuaVSy3trxy+FNUeRY78d+ICqmVkBueVuZqPWCd0U3aZhyV3STOB6YAzw9YhY0KhttSPvchZTt8d1rbot/tvxssANSe6SxgBfAz4C9ANPSFoaEc82YntmzeC4TrotcXeqRrXcpwNrIuJ5AEl3AucAXfUlsOHVM1E0oaVUqLiupbXpRL7rdvWzq1ccNyq5TwbWlkz3AyeXLiBpDjAnT26T9FyD6tJuJgKvtLoSTdaU96xrh5x9eB02MWxcQ2fH9jCf4XAc23Uwwv9B1bhuVHJXhbLYaSJiIbCwQdtvW5L6IqK31fVopgK952HjGhzbra5HM7Xze27UqZD9wNSS6SnAugZty6xZHNfWMRqV3J8Apkk6QtIewCxgaYO2ZdYsjmvrGA3plomI7ZLmAj8knTJ2S0SsbMS2OlDX7a5TkPfsuB5WIf7PI9S271kR7+oyNDOzDufLD9SZpOWSfiNpW348VzZ/hqTVkt6Q9JCkepzFUakOn6v3eq27SJorqU/SW5JuqzC/aiwruVbSpvz4G0mVDkiPpn49kkKSR9pX4OTeGHMjYlx+HDVYKGkisAT4EjAB6AO+3aI6mg1nHXANcEv5jBpieQ5wLnACcDxwNnBxg+trJZzcm+sTwErSYJingZnASZJOqrSwpAmSbpW0TtJmSffk8vGSlkkayOXLJE3J8+YDHwRuyHsON+TyoyXdL+lVSc9J+lTJdg6Q9D1JWyU9IekaSY+O9s1Kmpm3tUbSvNGuz5orIpZExD3ApgqzPwGsjIjvRMRvgKuBEyTdLGk18BXgDWBbRPw/4G+Bz1bblqR/LekfJW2RtFbSZ3P5WZKeyrG5VtLVJS97JP/dkmP9lPyaP5O0Kn83fli2R/HRHJOvSbpR0sO7spfbEbEdEX7U8QEsBwZIAxv+L3BaybzrgZuAjwJjc9kAcHeVdd1Hag2NB3YHPpzLDwDOA/YG9gW+A9xTVofPlUzvQxp886ekg+h/mOt3XJ5/Z37sDRybl310lJ/DGOCfgfcAewD/BBzb6v+PH7v0v7wGuK2s7HrgprKynwN/lWPsNeAbwLV5Xi/wepX1Hwa8DlyQ4/wA4MQ87zTgD0gN0eOBDcC5eV4PaZzB2JJ1nQusAY7J9fgvwD/meROBraQfprHAZcDvSr8rNX4eHRHbbrnX3xWkf/pk0pH070k6Ms8bB7wWET+KiO257BXg4PKVSDoE+BhwSURsjojfRcTDABGxKSLuiog3IuJ1YD7w4SHqdDbwYkTcGhHbI+JJ4C7gk/l6KecBV+X1PQssGuVnACVD9SPit6Qfj3PqsF5rD+NICbzUa8DzObbHAY+TxgIMzhtXpd/9T4AHIuKOHOebIuJpgIhYHhHPRMQ7EfEz4A6GjvWLgS9HxKpcj78GTsyt9zNJextL8ryvAr/ahffeEbHt5F5nEfFYRLweEW9FxCJS6/3MPHsbsF/ZSyaTvgTlpgKvRsTm8hmS9pb095JekrSVtHu6f07UlRwOnJx3ebdI2kL6Qh0MTCK1YkqH1a+tsI6RqjRUf3Id1mvtoVIs70dqgQ/O/7fAD0rmbYvc9C0zldQSfhdJJ+eDtQOSXgMuIbXAqzkcuL4kzl8ljSyeDBxKSUzmuvQPsa5qOiK2fZS5QSQ9QEqehwPHS7qY1L2ytWSZq0ldITdVWMVaYIKk/SNiS9m8y4GjgJMj4leSTgSeYsfw+PIv0Frg4Yj4SIV6jgG2k1pYv8jFU8uX2wU1DdW3zpNj+xhSg+JDg8XAvyIdUwLYQuoOvD1Pn1Ayr9xaUmu4km8BNwAfi4jfSLqOHcm9UjytBeZHxO3lMyRNY8eeBHkvYkr5cjXoiNh2y72OJO0v6QxJe5EOln6Z9Bn/cUS8DzgROFTSefkgzueBJyNidfm6ImI9qdVzYz6AunvJF2lf4E3SgaQJwFVlL99A6hoatAx4r6QL83p2l3SSpGMi4m3SWQ9X5z2Co4HP1OHj8FD9DidpbI7lMcAYSXtJGhsRp5NieTsp9npJMfZURKyWNBt4m9RwOVTSoaQGyW1VNnU7cLqkT+VtHpAbLJBi/dWc2KcD/67kdQPAO+wc638HXCnpuPwe/qWk8/O8+4A/kHRuPn3yUip0idagM2K71Z3+RXqQujieIO2abgF+CnykbJnTSa2Ld0hdNj1DrG8Cqf97A7AZWJLLDyUdNN1Gam1fTMmBJeCUXL4Z+GouO4oU3AOksx9+zI6DVpPyvK25/tcCD47ysxgLPA8cwY6DTse1+n/kx4j+h1fnuCp9XF0y/3RgNamhsZx0gHMm6RLIk4C/IXWLvJqfa4htfRB4LMfgWmB2Lv8k8FL+Ti0jteK/WfK6v8oxvQX4QC67EHimZF23lCw/M383XgNuBH4CXDjCz6UjYtsjVFtA0hpgT3acYvbTiLikhVXaiaRrgYMjYvYo13MmcB07hurPr0f9rH21e2yXkrQbqRX+JxHx0Ahf2/ax7eRu5K6YPUitnZOA75NOD7unpRUzqzNJZ5D2EN4E/jOpa+Y9EfFmSyvWAD6gapD6Ne8gdfdsJA04ubelNTJrjFNIB2n3IHUfnVvExA41tNzzAZVHSLtaY4HvRsRV+UDet0n9bC8Cn4p82p6kK4GLSAdV/mNE/LBRb8BsV0m6hTQGYGOkA96DZzB9ntSPC/CFiPh+nue4to5RS3IXsE9EbJO0O/AoaWTXJ0hHsRfk4bfjI+IKSceSWoHTSS3BB4D3Rjorw6xt5LOPtgHfKEvu2yLif5Qt67i2jjJst0yk7L8tT+6eH0EakXVaLl9EOlp+RS6/MyLeAl7IB1imk45KVzRx4sTo6enZpTdgVosVK1a8EhGTSssi4hFJPTWuYsRxDY5ta6xKcT2opj73PNBlBWmQwtci4jFJB0U6F5uIWC/pwLz4ZNIpgIMqjt5SyU2EDzvsMPr6+mp9P2YjJumlESw+V9JnSFc6vDx3N9YU13lbjm1riqHiuqZBTBHxdkScSDpZf7qk9w21vUqrqLDOhRHRGxG9kyZV/OExa4WbgCNJg3TWkw4uwwhGJTq2rR2MaIRqpGHwy0kDATbki1sNXuRqY16sM0ZvmVUQERtyY+Yd4H+xY1i849o6yrDJXdIkSfvn5/+CHaPSlgKDg1xms+PUuaXALEl7SjoCmEblC2OZtZ3BBkv2b0iXsQXHtXWYWvrcDwEW5X733YDFEbFM0k+AxZIuAl4GzgeIiJWSFpPOId0OXOozCoqjZ959v3/+4oKzWliT0ZN0B+mkgImS+knXSTktX9ckSKf4XgyO66IrUlwPquVsmZ8B769QvgmYUeU180nXGDdrWxFxQYXim4dY3nFtHcNXhTQzKyAndzOzAnJyNzMrICd3M7MCcnI3MysgJ3czswJycjczKyAndzOzAnJyNzMrICd3M7MCcnI3MysgJ3czsxI98+7b6UJincrJ3cysgJzczcwKyMndzKyAnNzNzArIyd3MuloRDp5W4uRuZlZATu5mZgXk5G5mVkBO7mZmBeTkbmZWQGNbXQFrf0U9m8CsyIZtuUuaKukhSaskrZR0WS6fIOl+Sb/Mf8eXvOZKSWskPSfpjEa+AWudTr8Gh6RbJG2U9POSMse1FUIt3TLbgcsj4hjgA8Clko4F5gEPRsQ04ME8TZ43CzgOmAncKGlMIypvNkq3kWK0lOPaCmHY5B4R6yPiyfz8dWAVMBk4B1iUF1sEnJufnwPcGRFvRcQLwBpger0rbjZaEfEI8GpZsePaCmFEB1Ql9QDvBx4DDoqI9ZB+AIAD82KTgbUlL+vPZeXrmiOpT1LfwMDAyGtu1hijimtwbFt7qDm5SxoH3AX8RURsHWrRCmXxroKIhRHRGxG9kyZNqrUaZq1SU1yDY9vaQ03JXdLupMR+e0QsycUbJB2S5x8CbMzl/cDUkpdPAdbVp7pmDee47iKdfELAcGo5W0bAzcCqiPhKyaylwOz8fDZwb0n5LEl7SjoCmAY8Xr8qmzWU49oKoZbz3E8FLgSekfR0LvsCsABYLOki4GXgfICIWClpMfAs6UybSyPi7brX3GyUJN0BnAZMlNQPXIXj2gpi2OQeEY9Sub8RYEaV18wH5o+iXmYNFxEXVJnluC64InfHDPLlB8zMCsjJ3cysgJzczcwq6PSuG184zCrq9MA263ZuuZuZFZCTu5lZATm5m5kVkJO7mXWVbjme5ORuZl2jWxI7OLmbmRWSk7uZWQE5uZuZFZCTu5lZATm5m5kVkJO7mVkBObmbmRWQLxxmO+mm84DNiswtdzOzAnJyNzMrICd3M7MCcnI3MysgJ3czK6Seefd19QkCTu5mZgXk5G5mVkDDJndJt0jaKOnnJWUTJN0v6Zf57/iSeVdKWiPpOUlnNKriZo0k6UVJz0h6WlJfLqsa91ZMndytU0vL/TZgZlnZPODBiJgGPJinkXQsMAs4Lr/mRklj6lZbs+b644g4MSJ683TFuLf2143978Mm94h4BHi1rPgcYFF+vgg4t6T8zoh4KyJeANYA0+tUV7NWqxb3Zm1nV/vcD4qI9QD574G5fDKwtmS5/lz2LpLmSOqT1DcwMLCL1TBrmAB+JGmFpDm5rFrc78Sxbe2g3teWUYWyqLRgRCwEFgL09vZWXMaao9t2V2t0akSsk3QgcL+k1bW+0LFt7WBXW+4bJB0CkP9uzOX9wNSS5aYA63a9ematERHr8t+NwN2k7sVqcW/WdnY1uS8FZufns4F7S8pnSdpT0hHANODx0VXRrLkk7SNp38HnwEeBn1M97q2NeE80GbZbRtIdwGnAREn9wFXAAmCxpIuAl4HzASJipaTFwLPAduDSiHi7QXW3NlH6ZXpxwVktrEndHATcLQnSd+RbEfEPkp6gQtybtaNhk3tEXFBl1owqy88H5o+mUmatFBHPAydUKN9Elbg3azceoWpmVkC+E1OXcr+kWbG55W5mVkBO7mZWaN26l+pumS7SrUFuxdcz776inKlVN07uZlYIpY2XejZkBtfVaT8e7pYxMysgJ3czswJycjczKyD3uZtZx/JJAtU5uRecg9+KpFMPbraCk3tBOambdTf3uZtZR3IDZmhO7mbWtqrd2NqJfXhO7mZmNei0HxT3uRdIpwWf2Ug4vkfGLXcza3tO7CPnlnuHa7egL+At98w6klvuZtY22q2x0snccu9A/gKY2XCc3M2s4UbSXdfOo1A7qdvRyb0NVWuZt3swleukL4I1T63x3Ql7qO38Q+Tk3gKVkl4tgdwJwV5NO38JrD667ce8/OYg7faefUDVzNpCJzde2lHDkrukmZKek7RG0rxGbcesmRzXyXCJuPSyAd2UtKtdLqEVFBH1X6k0BvgF8BGgH3gCuCAinq20fG9vb/T19dW9Hs20q10t3ax0N3a4XfrR7vJLWhERvSN+4c7rGFFcQ3vHdi1dCdU+9/LYfnHBWY73MtXywFDHFirNG+p/NFRcN6rPfTqwJiKezxW4EzgHqPolqKaWL/VwQTVUUNZiJEloV7fRjap9TiP5/Jrcz1u3uIbhv7jlxylKp6u9ttJryuN3qLgdSX0qzbMdhorvZjQAG9Vy/yQwMyI+l6cvBE6OiLkly8wB5uTJo4Dn6l6R9jQReKXVlWiydnjPh0fEpNGsoJa4zuWO7e7R6vdcNa4b1XJXhbKdfkUiYiGwsEHbb1uS+kbbPdBpCvSeh41rcGy3uh7N1M7vuVEHVPuBqSXTU4B1DdqWWbM4rq1jNCq5PwFMk3SEpD2AWcDSBm3LrFkc19YxGtItExHbJc0FfgiMAW6JiJWN2FYH6rrddQrynh3XwyrE/3mE2vY9N+SAqpmZtZZHqJqZFZCTu5lZATm5t4Ck/y5ptaSfSbpb0v6trlMjeKh+93Fstw8n99a4H3hfRBxPGs5+ZYvrU3d5qP7XgI8BxwIXSDq2tbWyJnBstwkn9xaIiB9FxPY8+VPS+dJF8/uh+hHxW2BwqL4VmGO7fTi5t96fAT9odSUaYDKwtmS6P5dZ93Bst5Bv1tEgkh4ADq4w64sRcW9e5ovAduD2ZtatSWoaqm+dx7HdGbHt5N4gEXH6UPMlzQbOBmZEMQcbeKh+QTm2OyO23S3TApJmAlcAH4+IN1pdnwbxUP0u5NhuHx6h2gKS1gB7Apty0U8j4pIWVqkhJJ0JXMeOofrzW1wlazDHdvtwcjczKyB3y5iZFZCTu5lZATm5m5kVkJO7mVkBObmbmRWQk7uZWQE5uZuZFdD/B0e3zpK/qpdcAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, axes = plt.subplots(nrows=2, ncols=2)\n",
    "axes[0,0].set(title='15 categ')\n",
    "axes[0,0].hist(data_df1,15)\n",
    "axes[0,1].set(title='30 categ')\n",
    "axes[0,1].hist(data_df1,30)\n",
    "axes[1,0].set(title='50 categ')\n",
    "axes[1,0].hist(data_df1,50)\n",
    "axes[1,1].set(title='100 categ')\n",
    "axes[1,1].hist(data_df1,100)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution théorique attendue\n",
    "Entre deux experiences, les quantités d'une protéine non affectée par le changement de conditions devraient-elles être strictement identiques ?\n",
    "\n",
    "Proposer une distribution théorique des **SA** des protéines **non affectées** par le changement de conditions.\n",
    " * Calculer les paramètres de cette distribution à partir de l'échantillon\n",
    " * Représenter ces paramètres par des lignes verticales sur l'histogramme précedent\n",
    "\n",
    "Implémenter une fonction qui prend un échantillon et **dessine** la distribution théorique et empirique. Elle utilisera la fonction ```empirical()```\n",
    "\n",
    "```python \n",
    "    def empiricalAndTheoric(data:pandas.Series)\n",
    "```\n",
    "\n",
    "##### Pour aller plus loin\n",
    "Vous dessinerez les histogrammes avec le paramètre `density=True` pour avoir une surface d'histogramme égale à 1. Ainsi l'aire de la fonction de densité étant égale à 1 les deux seront comparables sur le même graphique.\n",
    "Si vous souhaitez garde la même échelle il faudra appliquer un échelle à la densité de probabilité (largeur de l'histogramme * nombre d'observation)\n",
    "\n",
    "\n",
    "Exemple\n",
    "```python\n",
    "arr = np.random.randn(1000)\n",
    "\n",
    "plt.figure(1)\n",
    "result = plt.hist(arr)\n",
    "plt.xlim((min(arr), max(arr)))\n",
    "\n",
    "mean = np.mean(arr)\n",
    "variance = np.var(arr)\n",
    "sigma = np.sqrt(variance)\n",
    "x = np.linspace(min(arr), max(arr), 100)\n",
    "dx = result[1][1] - result[1][0]\n",
    "scale = len(arr)*dx\n",
    "plt.plot(x, norm.pdf(x, mean, sigma)*scale)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "print(result)\n",
    "print(result[1][1])\n",
    "print(result[1][0])\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.846424487\n"
     ]
    }
   ],
   "source": [
    "#extract mean/std , min/max from sample \n",
    "mean = np.mean(data_df1)\n",
    "sigma =np.sqrt(np.var(data_df1))\n",
    "mini= np.min(data_df1)\n",
    "maxi= np.max(data_df1)\n",
    "print(maxi)\n",
    "\n",
    "#sort x values from min to max\n",
    "df_sort = np.sort(data_df1)\n",
    "\n",
    "#calculate y=f(x)\n",
    "\n",
    "from scipy.stats import norm\n",
    "\n",
    "#def empiricalAndTheoric(data):\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Seuil de significativité\n",
    "Comment déterminer un seuil de *SA* au delà duquel seulement 5% des protéines **non affectées** seraient selectionnées ?\n",
    "\n",
    "Pour une Méthode de reconnaissance des protéines sur abondantes que siginifie ce taux de 5% ?\n",
    "\n",
    "Implémenter une fonction qui prend un échantillon et un pourcentage en argument et retourne la valeur seuil correspondant à cet argument.\n",
    "\n",
    "```python \n",
    "    def FDR_threshold(data:pandas.Series, t=0.05) -> Number# Où t peut valoir 5%, 2.5% ou 1%\n",
    "```\n",
    "\n",
    "Creer la fonction suivante\n",
    "\n",
    "```python\n",
    "def empiricalAndTheoricTreshold(data:pandas.Series, t=0.05)\n",
    "```\n",
    "\n",
    "Elle combinera les fonctions ```FDR_threshold``` et ```empiricalAndTheoric``` pour ajouter à son histogramme une ligne verticale rouge à l'emplacement du seuil.\n",
    "\n",
    "Aide: L'inverse de la fonction de densité cumulée (aka: [percent-point-function](https://en.wikipedia.org/wiki/Quantile_function)) de la loi normale est [implémentée dans scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html) (**ppf**).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optionnel\n",
    "##### Visualisation des seuils et de la distribution empirique\n",
    "Implémenter une fonction qui prend un échantillon et dessine la distribution théorique, empirique et colore [l'aire sous la courbe](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.fill_between.html) théorique au delà du seuil demandé.\n",
    "Vous pourrez ajouter les légendes correspondant à $\\mu, 2\\sigma, \\tau$ \n",
    "\n",
    "```python \n",
    "    def empiricalAndTheoricThresholdArea(SA_list, parameters*)\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pour aller plus loin\n",
    "\n",
    "###### La loi normale est un modèle contestable ici\n",
    "https://en.wikipedia.org/wiki/Cauchy_distribution\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Definitions de pathway fonctionels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nous disposons d'un jeu de protéines caractérisées par\n",
    "\n",
    "* Un ratio de quantification entre deux experiences\n",
    "* Une fiche uniprot \n",
    "\n",
    "Par ailleurs, l'ontologie GO est disponible via\n",
    "* un fichier plat de définition\n",
    "* une librarie pour la manipuler\n",
    "\n",
    "Il nous est demandé de detecter si certains termes GO sont fortement représentés parmi les protéines SurAbondantes.\n",
    "Par la suite, on utilisera indifférement les termes **Pathway** et **annotation GO**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation théorique de l'abondance\n",
    "\n",
    "* $Q+$ : Proteines Sur-représentées\n",
    "* $\\overline{Q+}$ : Proteines non Sur-représentées\n",
    "\n",
    "* $PW^{a}$ : Nombre de protéines membres du Pathway$a$\n",
    "* $\\overline{PW^{a}}$ : Nombre de protéines non-membres du Pathway$a$\n",
    "\n",
    "\n",
    "\n",
    " Proteines | $Q+$  | $\\overline{Q+}$ | Total lignes\n",
    " :--- | --- |--- |---|\n",
    "membres de $PW^{a}$  | $PW^{a}_{Q+}$ | $PW^{a}_{\\overline{Q+}}$ | $nPW^{a}$\n",
    "non-membres de $PW_{a}$ | $\\overline{PW^{a}}_{Q+}$ | $\\overline{PW^{a}}_{\\overline{Q+}}$ | $n\\overline{PW^{a}}$\n",
    "  **Total colonnes** | n$Q+$ | n$\\overline{Q+}$ | **nTOT**\n",
    " \n",
    " \n",
    "**Question**:\"Le nombre de protéines sur-représentées dans le Pathway$a$ est-il grand comparé à ce qui serait attendu au hasard?\"\n",
    "\n",
    "\n",
    "<!--## Deprecated \n",
    " **Question**:\"Les protéines sur-représentées sont-elles plus fréquentes dans le Pathway$a$ que dans les protéines non-membre du Pathway$X$?\"\n",
    " \n",
    "L'hypothèse nulle est \"Les protéines membres du pathway et non-membres du pathway ont des quantifications similaires\".\n",
    " -->\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Modélisation du hasard par une loi hypergéométrique\n",
    "\n",
    "Pour un pathway donné ($PWa$), chaque protéine sur-abondante peut être vue comme un tirage dans le pool de protéines de l'expérience. La population de ce pool est consituée de deux groupes:\n",
    " * Le groupe des protéines appartenant à $PWa$\n",
    " * Les autres ($\\overline{PW^{a}}$)\n",
    "\n",
    "La probabilité d'observer $k$ protéines de $PWa$ parmi les $n$ sur-abondantes se modélise alors naturellement par la loi de probabilité hypergéométrique.\n",
    "\n",
    "Rappellez-vous,\n",
    "\n",
    "Si X est la variable aléatoire du nombre de succès parmi n observations, sans remise.\n",
    "Alors la probabilité d'oberver exactement $k$ succès, suit une loi hypergéométrique de paramètres $N, n$ et $k$\n",
    "\n",
    "$P(X=k) = \\frac{ {K\\choose k} {N-K\\choose n-k} } { {N\\choose K} }$\n",
    "\n",
    "où\n",
    "\n",
    "\n",
    "* $K$ est le nombre de succès dans la population totale\n",
    "* $n$ est le nombre d'observations\n",
    "* $N$ est la taille de la population totale\n",
    "* $k$ est le nombre de succès parmi les observations\n",
    "\n",
    "Si **pour chaque $Pathway^{a}$**, $X$ est la variable aléatoire du **nombre de protéines sur-représentée** dans l'experience.\n",
    "\n",
    "\n",
    "Cocher les relations correspondantes\n",
    "\n",
    "\n",
    "  x  | $nTOT$ | $nPW^{a}_{Q+}$ | $nPW^{a}$ | $nQ+$\n",
    " --- | ---   | ---     | ---  | --- |\n",
    " $K$ |       |         |      | \n",
    " $n$ |       |         |      | \n",
    " $N$ |       |         |      | \n",
    " $k$ |       |         |      | \n",
    " \n",
    " \n",
    " \n",
    " \n",
    "Rappel: $nPW^{a}_{Q+}$ est le nombre de protéines appartenant au $Pathway^{a}$ parmi les protéines surexprimées\n",
    "\n",
    "Que représentent alors les deux expressions suivantes:\n",
    "\n",
    "$ P(X=nPW^{a}_{Q+})= \\frac{ {nPW^{a}\\choose X} {n\\overline{PW^{a}}\\choose nQ^{+} - X} } { {nTOT\\choose nPW^{a}} }$\n",
    "\n",
    "\n",
    "$P(X >= PW^{a}_{Q+})  = \\sum_{x_{i}=PW^{a}_{Q+}}^{x_{i} \\le nQ+}\\frac{ {nPW^{a}\\choose x_{i}} {n\\overline{PW^{a}}\\choose nQ^{+} - x_{i}} } { {nTOT\\choose nPW^{a}} }$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Créer la fonction qui retournera la probabilité, sous un [modèle hypergeométrique](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.hypergeom.html), d'obtenir $k$ succès ou plus,\n",
    "\n",
    "```python\n",
    "def righEnd_pValue(N, n, K, k) => p\n",
    "```\n",
    "Aux arguments:\n",
    "* $N$, le nombre total d'objet\n",
    "* $n$, le nombre de tirages\n",
    "* $K$, le nombre total d'objet de *typeI*\n",
    "* $k$, le nombre de tirage d'objet de *typeI* (ie: nombre de succès parmi les tirages)\n",
    "\n",
    "##### Dessiner les lollipop charts de la densité de probabilité et de la densité cumulée de la loi\n",
    "\n",
    "L'enrichissement en terme GO parmi les protéines surabondantes sera estimé par cette fonction, avec\n",
    "* $N$, le nombre total de protéines annotées dans le protéome\n",
    "* $n$, le nombre de protéines surabondantes\n",
    "* $K$, le nombre de protéines annotées par ce terme GO dans le protéome\n",
    "* $k$, le nombre de protéines annotées par ce terme GO dans les protéines surabondantes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prise en compte des annotations GO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Chargement des données uniprot et de l'ontologie GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprotCollection = uniprot.UniprotCollection(dataDir +\"/dataset\")\n",
    "K12 = uniprot.UniprotCollection(dataDir + \"/K12_proteome/\")\n",
    "print(\"Loading ontology\")\n",
    "go.setOntology(dataDir + \"/go.owl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manipulation de termes GO à l'aide de *go.py*\n",
    "\n",
    "Règle du vrai chemin: *Si une protéine est porteuse d'un terme GO, alors est elle aussi porteuse de tous les parents de ce terme.*\n",
    "\n",
    "#### AnnotationTree est l'arbre des annotations GO portées par une collection de proteines \n",
    "\n",
    "##### Un arbre ne rend compte que d'un seul namespace GO\n",
    "On définit le namespace des termes GO à extraire par un de ces trois termes : \n",
    "* \"biological process\", \"molecular function\", \"cellular component\"\n",
    " \n",
    "```python\n",
    "aTree = go.AnnotationTree(\"biological process\", collapse=True)\n",
    "```\n",
    "On ne garde dans cet arbre que les termes portés par les protéines données en dernier argument\n",
    "```python\n",
    "aTree.read_DAG(['P75936', 'P76231', 'P0A8S9'], uniprotCollection)\n",
    "```\n",
    "\n",
    "#### API de l'arbre d'annotations\n",
    "\n",
    "\n",
    "##### Taille de l'abre\n",
    "```python\n",
    "aTree.dimensions\n",
    "```\n",
    "\n",
    "##### Extraire un sous arbre\n",
    "```python\n",
    "bTree = aTree.newRoot(\"transmembrane transport\")\n",
    "```\n",
    "\n",
    "##### Enumerer les protéines porteuses de termes GO\n",
    "\n",
    "###### Dans tout l'arbre\n",
    "```python\n",
    "proteinList = aTree.getMembers()\n",
    "```\n",
    "###### A partir d'un noeud\n",
    "```python\n",
    "proteinList = aTree.getMembersByName(\"transmembrane transport\")\n",
    "proteinList = aTree.getMembersByID(\"GO:0055085\")\n",
    "```\n",
    "\n",
    "##### Acceder directement à un noeud\n",
    "```python\n",
    "goTerm = aTree.getByName(\"transmembrane transport\")\n",
    "goTerm = aTree.getByID(\"GO:0055085\")\n",
    "```\n",
    "\n",
    "##### Elager un arbre\n",
    "Retirer tous les noeuds n'ayant aucun des éléments suivants parmi leurs protéines annotées\n",
    "\n",
    "```python\n",
    "def predicat(goTerm):\n",
    "    return set(goTerm.getMembers()) & set(['P75936', 'P76231', 'P0A8S9'])\n",
    "\n",
    "bTree = aTree.drop(predicat)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluer la pValue d'enrichissment en terme GO d'une liste de protéine donnée\n",
    "\n",
    "Avant d'appeler `righEnd_pValue`, on devra donc\n",
    "\n",
    "1. Lister les protéines surabondantes\n",
    "2. Lister toutes les protéines sous le terme GO de votre choix \n",
    "    1. dans l'experience\n",
    "    2. dans le protéome\n",
    "3. Calculer l'intersection entre *A* et *B*\n",
    "\n",
    "###### *Aide:*\n",
    "Vous vous aiderez \n",
    "- de l'API de `go.annotationTree` \n",
    "- des commandes [`set` de python](https://docs.python.org/3/library/stdtypes.html#set)\n",
    "- du [filtrage de colonne Pandas](https://cmdlinetips.com/2018/02/how-to-subset-pandas-dataframe-based-on-values-of-a-column/) pour évaluer la liste de protéines surabondantes\n",
    "- de la [fonction subplot de matplotlib](https://stackoverflow.com/questions/42818361/how-to-make-two-plots-side-by-side-using-python), pour dessiner les deux courbes cote-à-cote\n",
    "\n",
    "Pour faciliter le travail nous prendrons le seuil de 0.0 pour detecter la surabondance d'une protéine"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyse de la surreprésentation\n",
    "\n",
    "Le package *stat_utils.py* permet d'appliquer cette analyse récursivement, en profondeur, à partir d'un terme GO général.\n",
    "\n",
    "#### Estimation récursive des enrichissements\n",
    "\n",
    "* Choix du terme GO général\n",
    "```python\n",
    "pathWayRoot = xpGoTree.getByName(\"transmembrane transport\")\n",
    "```\n",
    "\n",
    "* Définition du terme GO regroupant tout le protéome\n",
    "```python\n",
    "pathWayBKG = fullEcoliGoTree.getByName(\"biological process\")\n",
    "```\n",
    "\n",
    "* Calcul de l'enrichissement en termes GO successifs parmi les protéines surabondantes (ici, *saList*)\n",
    "```python\n",
    "oraFisher, oraCDF = computeORA_BKG(pathWayRoot, saList, pathWayBKG)\n",
    "```\n",
    "* Des arbres peuvent également être passés.\n",
    "```python\n",
    "tm_transport_GoTree = xpGoTree.newRoot(\"transmembrane transport\")\n",
    "oraFisher, oraCDF = computeORA_BKG(tm_transport_GoTree, saList, fullEcoliGoTree)\n",
    "```\n",
    "\n",
    "#### Affichage des enrichissements\n",
    "La fonction ci-dessous va vous permettre d'afficher les pathway par ordre croissant de p-value\n",
    "```python\n",
    "def printRankings(saList,_oraFisher, _oraCDF):\n",
    "    fisherRanking = [ ( d[0], d[1].name, len( set(d[1].getMembers()) & set(saList) ), len(saList) ) for d in sorted(oraFisher, key=lambda x:x[0]) ]\n",
    "    cdfRanking = [ ( d[0], d[1].name, len( set(d[1].getMembers()) & set(saList) ), len(saList) ) for d in sorted(oraCDF, key=lambda x:x[0]) ]\n",
    "    print(f\"FISHER test Rankings:\\n{fisherRanking}\")\n",
    "    print('\\n---\\n')\n",
    "    print(f\"CDF Rankings:\\n{cdfRanking}\")\n",
    "```\n",
    "\n",
    "##### Intégration des résultats à la visualisation du notebook \n",
    "\n",
    "En vous inspirant de la fonction ci-dessus, proposer une classe `Rankings` implémentant une méthode [`_repr_html_`](https://ipython.readthedocs.io/en/stable/config/integrating.html) qui permette un affichage plus lisible des résultats, comme par exemple:\n",
    "\n",
    "![caption](figs/_repr_html_.png)\n",
    "\n",
    "#### Appliquer cette analyse à terme GO de votre choix (et tous ses fils)\n",
    " - Il serait souhaitable d'encapsuler les différentes étapes dans une fonction de ce type\n",
    "```python\n",
    "    def computePathway(goTermString, xpTreeTerm, bkgTreeTerm, uniprotIdList) -> oraFisher, oraCDF\n",
    "```\n",
    "#### Appliquer cette analyse à tous les termes GO à un des arbres \"biological process\", \"molecular function\", \"cellular component\"\n",
    "\n",
    "***\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualisation des arbres\n",
    "Nous allons maintenant représenter la hiérarchie des termes GO portés par les protéines surreprésentées.\n",
    "Dans un premier temps, il vous est conseillé de vous placer à un niveau suffisamment bas dans l'arbre pour constuire la visualisation avec peu de feuilles.\n",
    "\n",
    "##### Elagage de l'arboresence GO\n",
    "Nous souhaitons, éliminer les branches de l'arbre ne comportant aucune protéine surreprésentées.\n",
    "On peut facilement obtenir toutes les protéines sous un terme GO donné, via l'API de l'arbre ou du terme GO.\n",
    "```python\n",
    "# Recupere depuis l'arbre toutes les occurences de protéines sous le noeud \"intracellular part\"\n",
    "proteinIdList = aTree.getMembersByNames(\"intracellular part\")\n",
    "# Recupere le noeud \"intracellular part\" depuis l'arbre, puis toutes les occurences de protéines sous celui-ci\n",
    "GOterm = aTree.getByNames(\"intracellular part\")\n",
    "_proteinIdList = GOTerm.getMembers()\n",
    "set(_proteinIdList) == set(proteinIdList) # -> True\n",
    "```\n",
    "\n",
    "L'API arbre fournit la méthode **drop**, qui prend en argument une fonction prédicat qui sera appliquée récursivement à chaque terme de l'arbre. \n",
    "* Le prédicat recoit un unique argument, le terme à évaluer.\n",
    "* Le prédicat doit retourner un booléen. \n",
    "* Si le prédicat retourne faux, le noeud évalué (ainsi que son sous-arbre) sera supprimé (\"droped\")\n",
    "\n",
    "### ete3\n",
    "\n",
    "La librairie [ete3](http://etetoolkit.org/) est une solution rapide pour déssiner des arbres en Python. \n",
    "Son intégration à [Jupyter](http://etetoolkit.org/ipython_notebook/) va vous permettre de dessiner la hiérarchie de termes GO portées par les protéines surreprésentées. Libre à vous d'essayer les différents [\"layout\"](http://etetoolkit.org/docs/latest/tutorial/tutorial_drawing.html#circular-tree-in-180-degrees)\n",
    "\n",
    "### amélioration\n",
    "\n",
    "Vous observerez que les arbres de termes GO des études experimentales ont été modifiés lors de l'analyse d'enrichissement. Chacun de leurs noeuds sont desormais porteurs:\n",
    "* d'une p-value d'enrichissement : `aTree.getByName(\"pore complex\").Fisher`\n",
    "* d'une liste de protéines effectivement annotées : `aTree.getByName(\"pore complex\").eTag`\n",
    "\n",
    "Les proteines annotées par ce terme ou ses descendants sont toujours accessibles: `set( aTree.getMembersByName(\"pore complex\") )`\n",
    "\n",
    "Vous pouvez rendre compte de ses valeurs en modifiant couleurs et tailles des noeuds correspondant dans la représentation graphique.\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualisation des abondances et Pathway pour chaque protéine\n",
    "Nous allons représenter la relation entre l'abondance des protéines (**Log2 Corrected Abundance Ratio**)  et la précision de la mesure (**LOG10 Adj.P-val**). Ceci se fera à l'aide de la commande matplotlibb [scatter](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.scatter.html).\n",
    "###### Volcano Plot global\n",
    "* Creer la fonction drawScatter(abundanceList, precisionList) qui produira le graphique.\n",
    "* Creer la fonction drawScatterPathway(goTermsList) \n",
    "    * Unique argument qui est une liste n goTerm aux p-value les plus faibles.\n",
    "    * Dessine les protéines annotées par ces termes GO\n",
    "    * Colorie de la même couleur, les protéines du même pathway\n",
    "    * Si une protéine appartient à différents pathway, la priorité ira à celui de p-value la plus faible.\n",
    "\n",
    "Au maximum 20 pathways pourront être représentés simultanément, vous utiliserez les [cartes de couleurs catégorielles](https://matplotlib.org/examples/color/colormaps_reference.html)\n",
    "\n",
    "######  Volcano Plot interactif\n",
    "\n",
    "Il serait plus intéressant de colorier les protéines d'un pathway donné à la volée, en mettant à jour le scatter plot à chaque fois. Jupyter fournit une libraire de composant interactif appelés [widgets](https://ipywidgets.readthedocs.io/en/stable/user_guide.html). \n",
    "\n",
    "* Construisez un widget `Dropdown` listant les pathway.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
